server/remote_cache/cachetools/cachetools.go

```diff
@@ -269,6 +276,232 @@ func UploadProtoToCAS(ctx context.Context, cache interfaces.Cache, instanceName
+	uploads          map[digest.Key]struct{}
+}
+
+func NewBatchCASUploader(ctx context.Context, env environment.Env, instanceName string) (*BatchCASUploader, error) {
+	bsClient := env.GetByteStreamClient()
+	if bsClient == nil {
+		return nil, status.InvalidArgumentError("Missing bytestream client")
+	}
+	casClient := env.GetContentAddressableStorageClient()
+	if casClient == nil {
+		return nil, status.InvalidArgumentError("Missing CAS client")
+	}
+	cache := env.GetCache()
+	if cache == nil {
+		return nil, status.InvalidArgumentError("Missing cache")
+	}
+
+	eg, ctx := errgroup.WithContext(ctx)
+	return &BatchCASUploader{
+		ctx:              ctx,
+		eg:               eg,
+		cache:            cache,
+		bsClient:         bsClient,
+		casClient:        casClient,
+		req:              &repb.BatchUpdateBlobsRequest{InstanceName: instanceName},
+		currentBatchSize: 0,
+		instanceName:     instanceName,
+		uploads:          make(map[digest.Key]struct{}),
+	}, nil
+}
+
+// Upload adds the given content to the current batch or begins a streaming
+// upload if it exceeds the maximum batch size. It closes r when it is no
+// longer needed.
+func (ul *BatchCASUploader) Upload(d *repb.Digest, r io.ReadSeekCloser) error {
+	// De-dupe uploads by digest.
+	dk := digest.NewKey(d)
+	if _, ok := ul.uploads[dk]; ok {
+		return r.Close()
+	}
+	ul.uploads[dk] = struct{}{}
+
+	r.Seek(0, 0)
+
+	if d.GetSizeBytes() > gRPCMaxSize {
+		ul.eg.Go(func() error {
+			_, err := UploadFromReader(ul.ctx, ul.bsClient, digest.NewInstanceNameDigest(d, ul.instanceName), r)
+			if err != nil {
+				return err
+			}
+			return r.Close()
+		})
+		return nil
+	}
+
+	if ul.currentBatchSize+d.GetSizeBytes() > gRPCMaxSize {
+		ul.flushCurrentBatch()
+	}
+	b, err := ioutil.ReadAll(r)
+	if err != nil {
+		return err
+	}
+	if err := r.Close(); err != nil {
+		return err
+	}
+	ul.req.Requests = append(ul.req.Requests, &repb.BatchUpdateBlobsRequest_Request{
+		Digest: d,
+		Data:   b,
+	})
+	ul.currentBatchSize += d.GetSizeBytes()
+	return nil
+}
+
+func (ul *BatchCASUploader) UploadProto(in proto.Message) (*repb.Digest, error) {
+	data, err := proto.Marshal(in)
+	if err != nil {
+		return nil, err
+	}
+	d, err := digest.Compute(bytes.NewReader(data))
+	if err != nil {
+		return nil, err
+	}
+	if err := ul.Upload(d, NewBytesReadSeekCloser(data)); err != nil {
+		return nil, err
+	}
+	return d, nil
+}
+
+func (ul *BatchCASUploader) UploadFile(path string) (*repb.Digest, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	d, err := digest.Compute(f)
+	if err != nil {
+		return nil, err
+	}
+	// Note: uploader.Upload will close the file.
+	if err := ul.Upload(d, f); err != nil {
+		return nil, err
+	}
+	return d, nil
+}
+
+type bytesReadSeekCloser struct {
+	io.ReadSeeker
+}
+
+func NewBytesReadSeekCloser(b []byte) io.ReadSeekCloser {
+	return &bytesReadSeekCloser{bytes.NewReader(b)}
+}
+func (*bytesReadSeekCloser) Close() error { return nil }
```

Comment:
I had hoped NopCloser would work here but doesn't look like it because of the seek.